### DISCUSSION TITLE
The memory usage of a program that repeatedly generates images gradually increases over time. Is this normal?

### DISCUSSION THREAD

**Discussion by C0j23T** [rust-skia/rust-skia#874] [Q&A] [answered]
I have a bot program that uses rust-skia to generate images. However, over time, I've noticed that the program's memory usage gradually increases from around 70MB(with some static resources) at the beginning to about 200MB after a few days. I've only been learning Rust for less than two months, and I'd like to know if this is normal.

Here is a code example:
```rust
let mut surface = surfaces::raster_n32_premul(ISize::new(1080, img_height)).expect("Cannot create surface");
let canvas = surface.canvas();
let mut paint = Paint::default();
paint.set_anti_alias(true);
// Some drawing code, such as:
let img = Image::from_encoded(Data::new_copy(&bytes));
canvas.draw_image(&img, Point::new(100.0, 100.0), Some(&paint));

paint.reset();
let image = surface.image_snapshot();
return match image.encode(None, EncodedImageFormat::JPEG, 50) {
    Some(data) => Some(data.as_bytes().to_vec()),
    None => {
        error!("Image generation failed");
        None
    }
};
// repeat this almost every 10 minutes
```

Translated by ChatGPT because my English is really bad😂

**Comment by tronical** [comment]
The symptom you’re describing doesn’t look normal to me. Do you have a standalone sample that reproduces the issue?

**Reply by C0j23T** [reply]
No, I don't have a standalone sample because I don't know which part is wrong😭, but I can give you part of my original code. It is kind of messy, sorry about that
https://gist.github.com/C0j23T/40c4711042f2c403778e75b199a90a7d

**Reply by tronical** [reply]
Thanks for the sample. I looked a bit through it but it's not obvious to me where it could leak :(. Are you sure that it's a leak without bounds or could it be that there is a ceiling?

Otherwise I suggest to run this with a leak detector like heaptrack.

**Reply by C0j23T** [reply]
I tried using heaptrack in my project, but I think it was interfered with by other parts, so I plan to set up a separate testing environment.<(。_。)>

In addition, I noticed that the examples of rust-skia usually only create a surface once. In my project, because the size of the images is uncertain each time, a new surface is created each time. Do you think the problem might be here?

**Reply by tronical** [reply]
I don’t think that might be the problem.

**Reply by C0j23T** [reply]
~~I modified the code so that it only used rust-skia, chrono,  once-cell and serde_json and then tested it with memcheck and massif from valgrind and heaptrack.~~

~~heaptrack said:~~
```
calls to allocation functions: 31223 (14665/s)
temporary memory allocations: 872 (409/s)
peak heap memory consumption: 33.93M
peak RSS (including heaptrack overhead): 48.33M
total memory leaked: 23.21M
```

~~valgrind:~~
~~https://gist.github.com/C0j23T/b3507334ab6ecbd711c3c88bfdebe9e6~~

~~(The second day)
I'm completely confused. In my bot program, the memory usage has clearly increased to almost 150MB, but the leak is still only around 23MB.
But without using it, the memory usage has always been around 30MB.~~

I tried conducting a comparative experiment and found that after several hours, the memory usage of sample that draws images is 60mb higher than sample that doesn't draw images. I think that the issue might lie in this.

**Comment by C0j23T** [comment]
@tronical This sample seems to have the same problem
```rust
use skia_safe::*;
use sysinfo::{System, SystemExt, ProcessExt};
use walkdir::WalkDir;

fn main() {
    const MB: f32 = 1048576.0;
    let mut sys = System::new();
    let pid = sysinfo::get_current_pid().unwrap();

    {
        sys.refresh_system();
        sys.refresh_process(pid);
        let process = sys.process(pid.clone()).unwrap();
        let prog = process.memory() as f32;
        println!("Initial memory: {:.2}MB", prog / MB);

        for entry in WalkDir::new("./imgs") { // example imgset
            let entry = entry.unwrap();
            if (&entry).file_type().is_dir() { continue; }
            let bytes = std::fs::read(&entry.path()).unwrap();
            println!("{:?}", entry);

            let img = Image::from_encoded(Data::new_copy(&bytes)).unwrap();
            let mut surface =
                surfaces::raster_n32_premul(ISize::new(1080, img.height())).expect("Cannot create surface");
            let canvas = surface.canvas();
            canvas.draw_image(
                &img,
                Point::new(0.0, 0.0),
                None,
            );
            let _output = surface.image_snapshot();
        }
    }
    sys.refresh_process(pid);
    let process = sys.process(pid.clone()).unwrap();
    let prog = process.memory() as f32;
    println!("Current memory: {:.2}MB", prog / MB);
}
```

It outputs
```
Initial memory: 5.25MB
DirEntry("./imgs/107882496_p0.png")
DirEntry("./imgs/104931251_p0.png")
DirEntry("./imgs/a_poetic_day_by_mauryart5_dfw1hrx.jpg")
DirEntry("./imgs/70022300_p0.png")
DirEntry("./imgs/107892370_p2.png")
DirEntry("./imgs/dragonstone_graveyard_by_hibelton_decx0um.jpg")
DirEntry("./imgs/75711345_p0.png")
DirEntry("./imgs/107049089_p0.png")
DirEntry("./imgs/107892370_p6.png")
DirEntry("./imgs/107892370_p13.png")
DirEntry("./imgs/107928311_p1.jpg")
DirEntry("./imgs/107887689_p0.png")
DirEntry("./imgs/107892370_p33.png")
DirEntry("./imgs/107892370_p3.png")
DirEntry("./imgs/107892370_p18.png")
DirEntry("./imgs/forgotten_kingdoms_iv_by_jjcanvas_dcozqaf.jpg")
DirEntry("./imgs/101582930_p16.jpg")
DirEntry("./imgs/107879681_p0.png")
DirEntry("./imgs/107887413_p0.png")
DirEntry("./imgs/61714645_p0.jpg")
DirEntry("./imgs/107892370_p20.png")
DirEntry("./imgs/107799979_p3.jpg")
DirEntry("./imgs/107778606_p0.png")
DirEntry("./imgs/107891135_p0.png")
DirEntry("./imgs/107892370_p14.png")
DirEntry("./imgs/107892370_p0.png")
DirEntry("./imgs/da8cxw4-85e55fa5-ebfc-47d3-a5c5-b5fd186a1558.png")
DirEntry("./imgs/107892370_p11.png")
DirEntry("./imgs/107913943_p0.png")
DirEntry("./imgs/107823999_p0.png")
DirEntry("./imgs/mouse_guard___owl_s_tree_by_darekzabrocki_ddwyd3r.jpg")
DirEntry("./imgs/ddunov8-93856880-85dc-4fa9-b9ba-35fc9ee0ea72.jpg")
DirEntry("./imgs/107892370_p8.png")
DirEntry("./imgs/107892370_p1.png")
DirEntry("./imgs/107904289_p0.png")
DirEntry("./imgs/107887732_p0.png")
DirEntry("./imgs/107890504_p0.png")
DirEntry("./imgs/107892370_p5.png")
DirEntry("./imgs/107892370_p26.png")
Current memory: 53.99MB
```

**Comment by C0j23T** [comment, marked as answer]
>Return an image backed by the encoded data, but attempt to defer decoding until the image is actually used/drawn. This deferral allows the system to cache the result, either on the CPU or on the GPU, depending on where the image is drawn. If memory is low, the cache may be purged, causing the next draw of the image to have to re-decode.

[from_encoded_with_alpha_type](https://rust-skia.github.io/doc/skia_safe/image/type.Image.html#method.from_encoded_with_alpha_type)

Images are cached. I didn't see this. 😂
